### ISSUE TITLE
confusing naming of cdr-atom and car-atom

### ISSUE THREAD

**Issue by noskill** [trueagi-io/hyperon-experimental#491]
Shouldn't we name these functions cdr-expr and car-expr since they work only with expressions?

**Comment by Necr0x0Der** [comment]
`cdr-expr` and `car-expr` might be a little bit more precise, but the difference is not critical in my opinion. If we want to change this, it should be done right away with properly informing metta coders. I'm ok with doing it if @vsbogd and @luketpeterson don't have objections. I'm also ok with leaving it as is otherwise.

**Comment by vsbogd** [comment]
Yes, I agree we could do it. It is better to release current version first to allow people using previous version by tag or from artifacts repository.

**Comment by luketpeterson** [comment]
Sorry to complicate things, but my personal opinion is that simply `car` and `cdr` are ok (the way minimal MeTTa was in the past) because they are such common operations and they have a very specific meaning in a whole family of languages.

But that's just my opinion so feel free to choose what you think is more in-line with the MeTTa philosophy.

**Comment by vsbogd** [comment]
`car` and `cdr` are too general and we may have to use them for the operations on `List` type instead of expressions.

**Comment by noskill** [comment]
If breaking existing code is not desirable there is an option to keep both names cdr-expr and cdr-atom for some time, but mark one of them as deprecated.

**Comment by noskill** [comment]
@vsbogd shouldn't it be possible to use different overloads of a function with different types like in c or namespaces?

**Comment by Necr0x0Der** [comment]
And I'm not sure in general if such overriding is actually a good idea or not...

**Comment by Necr0x0Der** [comment]
In any case, I would be ok to use `car` and `cdr` for Expressions only if we have this overloading. Otherwise we either stay with `-atom` or move to  more precise`-expr`.

**Comment by Necr0x0Der** [comment]
As for type-based dispatching, a clearer approach would be to have not
```
(: plus (-> Number Number Nubmer))
(: plus (-> Nat Nat Nat))
(= (plus $x $y) (+ $x $y))
(= (plus (S $x) $y) (plus $x (S $y)))
```
but something like
```
(= ((: plus Number) (: $x Number) (: $y Number)) (+ $x $y))
(= ((: plus Nat) (: $x Nat) (: $y Nat)) (plus $x (S $y)))
```
or another syntactic way, e.g.
```
(: plus (-> Number Number Number)) \
(= (plus $x $y) (+ $x $y))
```
to annotate types for particular expressions (see #177 ). In this case, it will not be precisely a non-deterministic type.

**Comment by Necr0x0Der** [comment]
Right now, we have a lot of other functions with `-atom` suffix (like `map-atom`), and it's already kind of conventional. Also, `car-expr` or `map-expr` sounds a little bit confusing as well, since they work on not on the whole expression, but only on the uppermost tuple, so `car-tuple` and `map-tuple` would be preferable. But `-atom` suffix underlines that these functions work directly on atomic meta-level representation rather on custom data structures. In any case, it seems a little bit late for renaming taking into account that there is no strong opinion on the necessity of such renaming. So, I'd personally keep going with `-atom` and close the issue if there are no strong objections.
